### PR TITLE
Adds support for adding package from remote urls

### LINKED ISSUE #695: Install of a package from URL
- [x] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.

## Issue

Sometimes packages easily avaliable only by url, for example [binaries for Windows](https://www.lfd.uci.edu/~gohlke/pythonlibs/). It would be nice to be able to do something like:

`poetry add pycurl --url https://download.lfd.uci.edu/pythonlibs/h2ufg7oq/pycurl-7.43.1-cp37-cp37m-win_amd64.whl`

For example, pipenv [supports it](https://github.com/pypa/pipenv/issues/480#issuecomment-330605268) :)

### PR DESCRIPTION
**Pull request by canassa**
- [ ] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

This PR adds support for installing packages from remote URLs, by passing a `--url` parameter. e.g.:

`poetry add minibar --url https://github.com/canassa/minibar/archive/master.zip`

I really needed this feature in order to keep using poetry in my internal company projects, therefore I hacked this PR together yesterday. This is really a proof-of-concept and not something production grade, I am opening this PR because I would like to get some feedback about this before I invest more time in it. Some question that I have are:

- Is this a welcomed feature?
- Do you agree with the new `--url` parameter, or do you prefer something else?
- Did I took the correct approach by subclassing `Dependency` and implementing the `search_for_url` that download the file and calls the `search_for_file` method?

Fixes #695 

Thanks!

### CODE REVIEW

**Comment by jaysonsantos** [comment]
I think you would have to implement your own dependency because if you use file's it would still reference your local file and if you distribute the lock you would have problems reproducing the install on a remote system for example.

Git dependency will pass all informations needed to reproduce on the remote system.
Here is an example of lock file using minibar with --git:

```
[[package]]
category = "main"
description = "Simple text progress bar library"
name = "minibar"
optional = false
python-versions = "*"
version = "0.4.0"

[package.source]
reference = "25484a383790ba09df69b4e1a6d65e018743326c"
type = "git"
url = "https://github.com/canassa/minibar"

```

**Comment by jaysonsantos** [comment]
Ah, git won't propagate hashes to metadata but IMO url should do it as file does

**Comment by canassa** [comment]
@jaysonsantos Thanks! I fixed the issue

**Comment by canassa** [comment]
I noticed that my implementation doesn't work for all kinds of files. For example, it works if I point the url to an archive that was generated with `poetry build` but it doesn't work if I try to point to a tarball with only the source code. For example, this works:

`poetry add minibar --url https://files.pythonhosted.org/packages/12/34/f1d8396d246f9649e164208d9c28e3c2bdb91077c59f86368b10fba36324/minibar-0.5.0.tar.gz`

But this doesn't work:

`poetry add minibar --url https://github.com/canassa/minibar/archive/0.4.0.tar.gz`

**Comment by canassa** [comment]
A friendly bump 😉 

I really would like to see this feature merge, please let me know if there is anything else I could do to help move this forward.

File: poetry/puzzle/provider.py [hunk 1]
```diff
@@ -198,6 +206,34 @@ def search_for_vcs(self, dependency):  # type: (VCSDependency) -> List[Package]
 
         return [package]
 
+    def search_for_url(self, dependency):  # type: (UrlDependency) -> List[Package]
+        tmp_dir = Path(mkdtemp(prefix="pypoetry-git-{}".format(dependency.name)))
```
Comment (romaia) [inline poetry/puzzle/provider.py]: maybe pypoetry-url-{}?

Comment (canassa) [reply poetry/puzzle/provider.py]: Thanks! fixed

**Comment by canassa** [comment]
I have been using this branch for a while now and so far so good!

The only problem that I had is that while this is working fine if I put the dependency in the `[tool.poetry.dependencies]` section, it doesn't seem to work if I put it in the `[tool.poetry.dev-dependencies]` section. I still didn't have the chance to debug that

**Comment by sklarsa** [comment]
What is the roadmap for getting this onto master?

**Comment by canassa** [comment]
@sklarsa  I think that it would at least require that the issue that I mentioned in my previous comment to be fixed. Other than that I think the PR works as I used my fork for quite a while without issues. Unfornutely, installing a poetry fork is a very cumbersome process and I don't want to force my coworkers into having to install a custom fork just to run my project.

Given that I haven't received any feedback from a maintainer in 2 months, I just assumed that this project is currently dead and I moved back to pip, therefore I won't be updating this PR any longer.

**Comment by drunkwcodes** [comment]
This feature overlaps with `git clone` and `wget`.

**Pull request closed without merging**